server.js

```diff
@@ -4160,6 +4161,268 @@ cache(function(data, match, sendBadge, request) {
   });
 }));
 
+// GitLab oauth integration
+camp.route(/^\/gitlab-auth$/, function(data, match, end, request) {
+  const baseURL = process.env.BASE_URL || 'https://img.shields.io';
+
+  if (!(serverSecrets && serverSecrets.gitlab_client_id)) {
+      return end('This server is missing GitLab OAuth secrets.');
+  }
+
+  const query = queryString.stringify({
+    client_id: serverSecrets.gitlab_client_id,
+    redirect_uri: (baseURL + '/gitlab-oauth/done'),
+    response_type: 'code',
+    // state: 'YOUR_UNIQUE_STATE_HASH' // FIXME implement CSRF tokens
+  });
+
+  const redirectURL = 'https://gitlab.com/oauth/authorize?' + queryString.stringify(query);
+
+  request.res.statusCode = 302;
+  request.res.setHeader('Location', redirectURL);
+  request.res.end();
+});
+
+// GitLab oauth callback integration
+camp.route(/^\/gitlab-auth\/done$/, function(data, match, end, request) {
+  const baseURL = process.env.BASE_URL || 'https://img.shields.io';
+
+  if (!(serverSecrets && serverSecrets.gitlab_client_id)) {
+      return end('This server is missing GitLab OAuth secrets.');
+  }
+
+  const options = {
+    method: 'POST',
+    json: true,
+    url: 'https://gitlab.com/oauth/token',
+    form: queryString.stringify({
+      client_id: serverSecrets.gitlab_client_id,
+      client_secret: serverSecrets.gitlab_client_secret,
+      code: data.code,
+      grant_type: 'authorization_code',
+      redirect_uri: (baseURL + '/gitlab-oauth/done')
+    })
+  };
+
+  // FIXME implement CSRF tokens
+
+  request(options, function(err, res, json) {
+    if (err != null) {
+      return end('The connection to GitLab failed.');
+    } else if (!json.access_token) {
+      return end('The GitLab OAuth process did not return an access token.');
+    }
+
+    gitlabHelpers.addToken(json.access_token);
+  });
+});
+
+// GitLab stars & forks integration
+camp.route(/^\/gitlab\/((http(?:s?))\/([^/]+)\/)?(star|fork)s\/([^/]+)\/(.+)\.(svg|png|gif|jpg|json)$/,
+cache(function(data, match, sendBadge, request) {
+  const isStar = (match[4] === 'star');
+  const user = match[5];
+  const repo = match[6];
+  const format = match[7];
+  const apiURL = gitlabHelpers.baseURL(match) + "/api/v4/projects/" + user + '%2F' + repo;
```

Comment:
Could we use either template string or all single quotes `'` here:
```javascript
const apiURL = `${gitlabHelpers.baseURL(match)}/api/v4/projects/${user}%2F${repo}`;
const apiURL = gitlabHelpers.baseURL(match) + '/api/v4/projects/' + user + '%2F' + repo;
```

Reply:
Besides consistency, is there a low-level reason behind keeping string literal types consistent?

Reply:
Just for consistency, really. Template string is the way forward.